### ISSUE TITLE
Enums inside of tagged, structural enums do not deserialize properly

### ISSUE THREAD

**Issue by JeffBelgum** [serde-rs/json#248]
I'm not sure whether this is a serde or serde_json issue. Hopefully I'm filing it in the right repo. 

The following test case shows that serialization works fine but deserialization of the serialized output doesn't succeed:

```
extern crate serde;
#[macro_use]
extern crate serde_derive;
extern crate serde_json;

#[derive(Serialize, Deserialize)]
#[serde(tag = "variant")]
enum A {
    A1,
    A2 { b: B },
}

#[derive(Serialize, Deserialize)]
enum B {
    B1,
}

fn main() {
    let a_enum = A::A2 { b: B::B1 };
    let a_json = r#"{"variant":"A2","b":"B1"}"#;
    assert_eq!(a_json, serde_json::to_string(&a_enum).unwrap());
    let a_deserialized: A = serde_json::from_str(a_json).expect("This should work");
}
```

Note that removing the `#[serde(tag = "variant")]` attribute and adjusting the serialized string appropriately does succeed.

**Comment by dtolnay** [comment]
@46bit is working on this over in https://github.com/serde-rs/serde/issues/775 so let's track it there.

**Comment by dtolnay** [comment]
This is fixed in serde 0.9.9.
